src/loc_decoder.rs

```diff
@@ -816,6 +816,19 @@ impl Expr {
                 }
                 _ => panic!("SeqLength: expected Seq"),
             },
+            Expr::SeqIx(seq, index) => {
+                match seq
+                    .eval_with_loc(scope)
+                    .coerce_mapped_value()
+                    .get_sequence()
+                {
+                    Some(values) => {
+                        let index = index.eval_value_with_loc(scope).unwrap_usize();
+                        Cow::Owned(values[index].clone())
```

Comment:
Can this return a reference to the value or is it not possible here?

Reply:
I do suppose it could, in point of fact. I will double-check that it works before I push that change.

Reply:
Ah, no, it actually can't be referenced. Because the capture of `values` lives behind the stack-local evaluation of `.eval_with_loc` which may be an owned value due to the `Cow` and therefore will not live long enough to be referenced. There may be some pattern-matches we could apply to possibly obtain a longer-lived reference to introspect and return, but that is arguably more work than is really worth bothering with...

Reply:
It is actually possible, but only if we do this:

```rust
            Expr::SeqIx(seq, index) => {
                let eval = seq.eval_with_loc(scope);
                match eval {
                    Cow::Borrowed(v) => match v.coerce_mapped_value().get_sequence() {
                        Some(values) => {
                            let index = index.eval_value_with_loc(scope).unwrap_usize();
                            Cow::Borrowed(&values[index])
                        }
                        _ => panic!("SeqIx: expected Seq"),
                    }
                    Cow::Owned(v) => match v.coerce_mapped_value().get_sequence() {
                        Some(values) => {
                            let index = index.eval_value_with_loc(scope).unwrap_usize();
                            Cow::Owned(values[index].clone())
                        }
                        _ => panic!("SeqIx: expected Seq"),
                    }
                }
            }
```

Reply:
We could clean it up, e.g. factoring out the duplicate evaluation of `index`, and maybe even find a way of producing this sort of thing auto-magically using either a helper function or a macro (though the latter might be a bit heavy-handed for this sort of one-off thing)

Reply:
Probably not worth overthinking, just wanted to check!

Reply:
I think I managed to finesse it a bit:

```rust
            Expr::SeqIx(seq, index) => {
                cow_map(seq.eval_with_loc(scope), |v| {
                    match v.coerce_mapped_value().get_sequence() {
                        Some(values) => {
                            let index = index.eval_value_with_loc(scope).unwrap_usize();
                            &values[index]
                        }
                        _ => panic!("SeqIx: expected Seq"),
                    }
                })
            }
```

with local function

```rust
fn cow_map<'a, T, U>(x: Cow<'a, T>, f: impl for<'i> Fn(&'i T) -> &'i U) -> Cow<'a, U>
where
    T: 'static + Clone,
    U: 'static + Clone + ToOwned,
{
    match x {
        Cow::Borrowed(x) => Cow::Borrowed(f(x)),
        Cow::Owned(x) => Cow::Owned(f(&x).to_owned()),
    }
}
```

Reply:
neat! almost feels like something like that should be in the standard library 🤔

Reply:
Also noticed that this same helper function can be used on `RecordProj` and `TupleProj`, as well as `SeqIx` in both `decoder.rs` and `loc_decoder.rs`.